internal/pkg/instrumentation/probe/probe.go

```diff
@@ -41,4 +55,254 @@ type Probe interface {
+	spec, err := i.SpecFn()
+	if err != nil {
+		return err
+	}
+
+	err = i.injectConsts(td, spec)
+	if err != nil {
+		return err
+	}
+
+	obj, err := i.buildObj(exec, td, spec)
+	if err != nil {
+		return err
+	}
+
+	i.reader, err = i.ReaderFn(*obj)
+	if err != nil {
+		return err
+	}
+	i.closers = append(i.closers, i.reader)
+
+	return nil
+}
+
+func (i *Base[BPFObj, BPFEvent]) injectConsts(td *process.TargetDetails, spec *ebpf.CollectionSpec) error {
+	opts, err := consts(i.Consts).injectOpts(td)
+	if err != nil {
+		return err
+	}
+	return inject.Constants(spec, opts...)
+}
+
+func (i *Base[BPFObj, BPFEvent]) buildObj(exec *link.Executable, td *process.TargetDetails, spec *ebpf.CollectionSpec) (*BPFObj, error) {
+	obj := new(BPFObj)
+	if c, ok := ((interface{})(obj)).(io.Closer); ok {
+		i.closers = append(i.closers, c)
+	}
+
+	sOpts := &ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{
+			PinPath: bpffs.PathForTargetApplication(td),
+		},
+	}
+	err := utils.LoadEBPFObjects(spec, obj, sOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	for symb, f := range i.Uprobes {
+		links, err := f(symb, exec, td, obj)
+		if err != nil {
+			return nil, err
+		}
+		for _, l := range links {
+			i.closers = append(i.closers, l)
+		}
+	}
+
+	return obj, nil
+}
+
+// Run runs the events processing loop.
+func (i *Base[BPFObj, BPFEvent]) Run(dest chan<- *Event) {
+	for {
+		record, err := i.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				return
+			}
+			i.Logger.Error(err, "error reading from perf reader")
+			continue
+		}
+
+		if record.LostSamples != 0 {
+			i.Logger.Info("perf event ring buffer full", "dropped", record.LostSamples)
+			continue
+		}
+
+		e, err := i.processRecord(record)
+		if err != nil {
+			i.Logger.Error(err, "failed to process perf record")
+		}
+
+		dest <- e
+	}
+}
+
+func (i *Base[BPFObj, BPFEvent]) processRecord(record perf.Record) (*Event, error) {
+	buf := bytes.NewBuffer(record.RawSample)
+
+	var event BPFEvent
+	err := binary.Read(buf, binary.LittleEndian, &event)
+	if err != nil {
+		return nil, err
+	}
+	return i.ProcessFn(&event), nil
+}
+
+// Close stops the Probe.
+func (i *Base[BPFObj, BPFEvent]) Close() {
+	var err error
+	for _, c := range i.closers {
+		err = errors.Join(err, c.Close())
+	}
+	if err != nil {
+		i.Logger.Error(err, "failed to cleanup", "Probe", i.Name)
+	}
+}
+
+// UprobeFunc is a function that will attach a eBPF program to a perf event
+// that fires when the given symbol starts executing in exec.
+//
+// It is expected the symbol belongs to are shared library and its offset can
+// be determined using target.
+//
+// Losing the reference to the resulting Link (up) will close the Uprobe and
+// prevent further execution of prog. The Link must be Closed during program
+// shutdown to avoid leaking system resources.
+type UprobeFunc[BPFObj any] func(symbol string, exec *link.Executable, target *process.TargetDetails, obj *BPFObj) ([]link.Link, error)
```

Comment:
The Uprobe attaching functions are very similar between all the probes (some may have return probes while others don't), but maybe we can have a generic function for loading the uprobes defined in here?

Reply:
The hard part is going to be that each function needs to access a particular field of the `BPFObj`. We could also add an additional interface that each probe could provide that would access the correct field, but that seems to also be about equivalent to having each probe provide an implementation for this.

I think we can explore options though in a future PR after this is merged.